addons/udes_security/controllers/main.py

```diff
@@ -0,0 +1,135 @@
+# -*- coding: utf-8 -*-
+
+import json
+import logging
+
+from odoo import http, SUPERUSER_ID, _
+from odoo.http import request
+from odoo.exceptions import UserError
+
+from odoo.addons.web.controllers.main import Binary, serialize_exception
+
+_logger = logging.getLogger(__name__)
+
+
+class BinaryExtension(Binary):
+    def _get_file_type(self, filename):
+        """Extract the file type from the filename by returning value
+           after the last '.' character"""
+        return filename.split(".")[-1].lower()
+
+    @http.route(
+        [
+            "/web/content",
+            "/web/content/<string:xmlid>",
+            "/web/content/<string:xmlid>/<string:filename>",
+            "/web/content/<int:id>",
+            "/web/content/<int:id>/<string:filename>",
+            "/web/content/<int:id>-<string:unique>",
+            "/web/content/<int:id>-<string:unique>/<string:filename>",
+            "/web/content/<string:model>/<int:id>/<string:field>",
+            "/web/content/<string:model>/<int:id>/<string:field>/<string:filename>",
+        ],
+        type="http",
+        auth="public",
+    )
+    def content_common(
+        self,
+        xmlid=None,
+        model="ir.attachment",
+        id=None,
+        field="datas",
+        filename=None,
+        filename_field="datas_fname",
+        unique=None,
+        mimetype=None,
+        download=None,
+        data=None,
+        token=None,
+        access_token=None,
+        **kw
+    ):
+        BlockedFileType = request.env["udes.blocked_file_type"].sudo()
+
+        if request.session.uid != SUPERUSER_ID:
+            # Check if the file type is blocked
+            if download:
+                file_type = ""
+                record_filename = ""
+
+                if filename:
+                    file_type = self._get_file_type(filename)
+                elif id and model and filename_field:
+                    RecordModel = request.env[model].sudo()
+                    record = RecordModel.browse(id)
+                    record_filename = record[filename_field]
+                    file_type = self._get_file_type(record_filename)
+
+                blocked_file_type_count = BlockedFileType.search_count([("name", "=", file_type)])
+
+                # If the user is trying to download a blocked file type then 
+                # prevent download and return an error message
+                if blocked_file_type_count:
+                    _logger.info(
+                        "Blocked attempt to download file %s" % (filename or record_filename)
```

Comment:
I think it could be useful to know the user too.

Can we unify `filename` and `record_filename` ?

Reply:
The user is informed below when ```download_error``` is returned which products a notification in the web client. If we raise an exception here the user will just be given a 500 error.

Reply:
Haven't explained properly :) I mean also add the user that tried to download:
```Blocked attemp to download file XXX by user YYY```

Reply:
Gotcha. Think that is more down to me reading too fast!

Good idea, will go add this. :)

Reply:
Have now created a ```_log_user_file_action_blocked``` function to handle this, and that includes the user login in the message.

Have also unified ```filename``` and ```record_filename```.